### PR TITLE
Add http2 status code in error_message if it's not 200

### PR DESCRIPTION
**Pull request by y-zeng**
If a stream is cancelled because the status code in response's http2 header is not 200, the status code will be put in the error_message

Related issue: #4168

### CODE REVIEW

**Comment by yang-g** [comment]
LGTM

**Comment by ctiller** [comment]
I don't see a test here.

Suggest adding something to at least the bad_client suite.

**Comment by y-zeng** [comment]
Thanks for the review! I'm working on the test.

**Comment by y-zeng** [comment]
Added bad_server_response_test. It tests the client behavior after receiving bad responses from the server.

File: test/core/end2end/bad_server_response_test.c [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+  return gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                      gpr_time_from_seconds(seconds, GPR_TIMESPAN));
+}
+
+static void start_rpc(int target_port, grpc_status_code expected_status,
+                      const char *expected_detail) {
+  grpc_op ops[6];
+  grpc_op *op;
+  grpc_metadata_array initial_metadata_recv;
+  grpc_metadata_array trailing_metadata_recv;
```
Comment (yang-g) [inline test/core/end2end/bad_server_response_test.c]: Likely you need to memset ops to quiet msan and call grpc_metadata_array_destroy() on the initial/trailing metadata_recv.

Comment (y-zeng) [inline test/core/end2end/bad_server_response_test.c]: Fixed. Thanks a lot for the review!

**Comment by yang-g** [comment]
LGTM
Please resolve the clang-format error:
--- /local-code/test/core/end2end/bad_server_response_test.c	2016-06-27 22:27:50.167420573 +0000
+++ /tmp/tmp.EzAEq2vsmQ	2016-06-27 22:39:04.473416359 +0000
@@ -126,8 +126,7 @@
     gpr_free(dump);
   }
 
-  gpr_log(GPR_DEBUG,
-          "got %" PRIuPTR " bytes, expected %" PRIuPTR " bytes",
+  gpr_log(GPR_DEBUG, "got %" PRIuPTR " bytes, expected %" PRIuPTR " bytes",
           state.incoming_data_length, EXPECTED_INCOMING_DATA_LENGTH);
   if (state.incoming_data_length > EXPECTED_INCOMING_DATA_LENGTH) {
     handle_write(exec_ctx);
Command exited with non-zero status 1

**Comment by yang-g** [comment]
LGTM

**Comment by jtattermusch** [comment]
needs resolving.

**Comment by y-zeng** [comment]
Resolved.

**Pull request merged**